src/Core/Domain/Product/ValueObject/RedirectionPage/CategoryRedirectionPage.php

```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\RedirectionPage;
+
+use PrestaShop\PrestaShop\Core\Domain\Category\Exception\CategoryException;
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryId;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\RedirectionPage\TypedRedirectionPageInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\RedirectionPage\ResponseCode;
+
+/**
+ * Category redirection page
+ */
+final class CategoryRedirectionPage implements TypedRedirectionPageInterface
```

Comment:
For me it seems more complex than it should be. :confused: 

What if we had VO like `OfflineRedirection` (name can be improved):

```php
class OfflineRedirection {
    /** @param string */
    private $redirectType; // value could be "temporary_to_product" or "permanent_to_product"

    /** @param int|null */
    private $resourceId; // Product or category id, or null if redirect type is "none"
}
```

Reply:
im not sure about this now - my suggestion avoids null values . Tried to achieve that by applying special case design pattern usage here.

Reply:
Yeah, but that approach seems to introduce a lot of complexity imo.